lib/sidekiq/xaclient.rb

```diff
@@ -0,0 +1,52 @@
+require "sidekiq/client"
+
+begin
+  require "after_commit_everywhere"
+rescue LoadError
+  Sidekiq.logger.error("You need to add after_commit_everywhere to your Gemfile for this to work")
+  exit(-127)
+end
+
+module Sidekiq
+  module XAClient
+    include AfterCommitEverywhere
+
+    ##
+    # Control job push within ActiveRecord transactions. Jobs can specify an
+    # "xa" attribute to define the policy to use:
+    #   * true or "commit" means enqueue the job after committing the current transaction.
+    #   * "rollback" means enqueue this job only if the current transaction rolls back
+    #   * nil or false means enqueue the job immediately, Sidekiq's default behavior
+    def push(item)
+      # Sidekiq::Job does not merge sidekiq_options so we need to fallback
+      policy = item.fetch("xa") { |key|
+        kl = item["class"]
+        kl.respond_to?(:get_sidekiq_options) ? kl.get_sidekiq_options[key] : nil
+      }
+      if policy == "commit" || policy == true
+        after_commit { super }
+      elsif policy == "rollback"
+        after_rollback { super }
+      else # enqueue immediately
+        super
+      end
+    end
+
+    ##
+    # We don't provide transactionality for push_bulk because we don't want
+    # to hold potentially hundreds of thousands of job records in memory due to
+    # a long running enqueue process. TODO: wdyt?
+    def push_bulk(items)
+      super
+    end
+  end
+end
+
+##
+# Use `Sidekiq.transactional_push!` in your sidekiq.rb initializer
+module Sidekiq
+  def self.transactional_push!(policy: "commit") # TODO: is this knob really necessary?
+    Sidekiq::Client.prepend(Sidekiq::XAClient)
```

Comment:
What's the advantage of using multiple inheritance (including modules in Ruby is multiple inheritance) over composition from #5248?

Reply:
The advantage is that we can mix and match with other Modules. With composition, I have to instantiate some known type with known behavior. With `prepend`, everyone instantiates the same concrete type with the Modules dynamically mixed in. Sidekiq Pro has two separate Modules which mix into `raw_push` to provide reliability and atomic push.